File: Patterns/Command Pattern/CSharp/Command/CailingFan/CailingFanOffCommand.cs
namespace Command.CailingFan
{
    public class CailingFanOffCommand : ICommand
    {
        private CailingFan cailingFan;
        private int preview;

        public CailingFanOffCommand(CailingFan cailingFan)
        {
            preview = CailingFan.OFF;
            this.cailingFan = cailingFan;
        }

        public void Execute()
        {
            preview = cailingFan.State;
            cailingFan.Off();
        }

        public void Undo()
        {
            switch (preview)
            {
                case CailingFan.HIGH:
                    cailingFan.SetHigh();
                    break;
                case CailingFan.MEDIUM:
                    cailingFan.SetHigh();
                    break;
                case CailingFan.LOW:
                    cailingFan.SetHigh();
                    break;
            }
        }
    }
}

File: Patterns/Command Pattern/Cpp/Command/CailingFan/CailingFanOffCommand.cpp
#include "../ICommand.cpp"
#include "./CailingFan.cpp"

class CailingFanOffCommand : public ICommand {
private:
    CailingFan* cailingFan;
    int preview;

public:
    CailingFanOffCommand(CailingFan* cailingFan) : cailingFan{cailingFan}, preview{cailingFan->OFF} {}

    void execute() override {
        preview = cailingFan->getState();
        cailingFan->off();
    }

    void undo() override {
        if (preview == cailingFan->HIGH) cailingFan->setHigh();
        else if (preview == cailingFan->MEDIUM) cailingFan->setMedium();
        else if (preview == cailingFan->LOW) cailingFan->setLow();
    }

    string getCommandName() override {
        return "CailingFanOffCommand";
    }
};